### DISCUSSION TITLE
Java Access Bridge calls ideally should not be done on NVDA main thread

### DISCUSSION THREAD

**Discussion by thgcode** [nvaccess/nvda#14461] [General]
### Is your feature request related to a problem? Please describe.
Yes. Some time ago I posted this snippet of code, that can freeze NVDA because Java Access Bridge takes some time to return the accessible context to NVDA. If the sleep interval is increased, NVDA can freeze as much as the time that the Java Access Bridge takes to return:

```
import javax.accessibility.AccessibleContext;
import javax.swing.*;

public class Main {
    public static void main(String[] args) throws Exception {
        var frame = new JFrame();
        frame.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);

        JButton start = new JButton("Start");
        frame.add(start);

        JButton quit = new JButton("Quit") {
            @Override
            public AccessibleContext getAccessibleContext() {
                return new JButton.AccessibleJButton() {
                    @Override
                    public String getAccessibleName() {
                        try {
                            Thread.sleep(2000);
                        } catch (InterruptedException e) {
                            throw new RuntimeException(e);
                        }

                        return super.getAccessibleName();
                    }
                };
            }
        };

        frame.add(quit);

        frame.setVisible(true);
    }
}
```

### Describe the solution you'd like
The idea would be to transfer the Java Access Bridge calls to another thread, and only post the event on the NVDA main thread when all accessible context and other needed information for the event is retrieved. This way, if Java freezes, only the thread responsible for doing the calls would be frozen.

### Describe alternatives you've considered
The alternative on this program would be to remove the sleep call and unblock Java Access Bridge as soon as possible, however it depends on a program update to do it.

### Additional context
NVDA calls the Java Access Bridge through the windowsAccessBridge32.dll library. This DLL uses the same thread as the program that called it (on NVDA case the main thread). If this DLL run some blocking code, NVDA would block as long as the time takes to the DLL to return.

**Comment by michaelDCurran** [comment]
This type of issue is true for pretty much all accessibility frameworks, including MSAA and UI Automation. NNVDA is primarily coded in a synchronous way from the ground up, and we have no plans right now to do a major redesign to switch to some way of making all calls asynchronous.
Even if this particular call was asynchronous, NVDA's main thread would still need to block at some point to wait for that call to complete. It is asking for the accessible name, it must block until it get it.
Certainly happy to have some philosophical discussion on future designs for screen readers, but right now this is where we are. And so far, this has not been a major issue.

**Reply by thgcode** [reply]
Hello,
Sorry for the delay, I am still investingating this issue on my spare time and I think I found some new information about it.
To give some more context, I created this artificial example to illustrate, but I am having an issue when IntelliJ receives large quantities of data from the console of the application it is running, in some machines it freezes NVDA and because Java Access Bridge makes a lot of sendMessage calls, in some cases NVDA shows an "Access Denied" error when trying to restart and I cannot use the machine because the keyboard hook is frozen as well (Iam not sure of this but I  think). But I found that the main issue is that for some reason, NVDA cannot cancel Java Access Bridge calls, even though it uses SendMessageTimeout. I am still investingating why. I created this discussion earlier because I did not knew about the cancel mechanism. For some calls I measured for example 2 seconds of the call being made and NVDA getting the result. I will post if I find new answers about this problem.

**Reply by thgcode** [reply]
I think I found the cause of NVDA not managing to cancel Java Access Bridge calls. The Windows Access Bridge DLL (at least the one NVDA uses), is compiled without UNICODE defined. So NVDA calls the real SendMessageTimeoutA function, bypassing the cancel mechanism. I am not sure what the correct solution would be, should we also create a hook for SendMessageTimeoutA, or compile the a new DLL that calls the SendMessageTimeoutW function?

**Comment by michaelDCurran** [comment]
I think it definitely makes sense to also hook the sendMessageA 
functions if we don't already.
Note that we don't compile windowsaccessbridge.dll ourselves, it is 
taken directly from a standard Java Runtime.

**Comment by seanbudd** [comment]
Closed by #19994
